File: main/client/src/mill/main/client/lock/MemoryLock.java
package mill.main.client.lock;

import java.util.concurrent.locks.ReentrantLock;

class MemoryLock extends Lock {

    private ReentrantLock innerLock = new ReentrantLock();

    public boolean probe() {
        return !innerLock.isLocked();
    }

    public Locked lock() {
        innerLock.lock();
        return new MemoryLocked(innerLock);
    }

    public Locked tryLock() {
        if (innerLock.tryLock()) return new MemoryLocked(innerLock);
        else return null;
    }

    @Override
    public void close() throws Exception {
        innerLock.unlock();
    }
}

File: main/client/src/mill/main/client/lock/MemoryLock.scala
package mill.main.client.lock

import java.util.concurrent.locks.ReentrantLock

class MemoryLock extends Lock {

  private val innerLock = new ReentrantLock()

  override def probe(): Boolean = {
    !innerLock.isLocked
  }

  override def lock(): Locked = {
    innerLock.lock()
    new MemoryLocked(innerLock)
  }

  def tryLock(): MemoryTryLocked = {
    if (innerLock.tryLock()) new MemoryTryLocked(innerLock)
    else new MemoryTryLocked(null)
  }

  @throws[Exception]
  override def close(): Unit = {
    innerLock.unlock()
  }
}
